spec/src/main/asciidoc/core/invokers.asciidoc

```diff
@@ -0,0 +1,181 @@
+
+When the target bean is not a managed bean, attempting to build an invoker leads to a deployment problem.
+When the target bean is an interceptor, attempting to build an invoker leads to a deployment problem.
+
+Multiple managed beans may inherit a method from a common supertype.
+In that case, an invoker must be built for each target bean individually.
+An invoker built for one target bean may not be used to invoke the target method on an instance of another target bean.
+
+The only way to build an invoker is using the `InvokerBuilder`.
+An `InvokerBuilder` can only be obtained in CDI portable extensions and build compatible extensions.
+See <<invoker_builder>> for more information.
+
+[[using_invoker]]
+=== Using an `Invoker`
+
+The `Invoker` interface contains a single method:
+
+[source,java]
+----
+public interface Invoker<T, R> {
+    R invoke(T instance, Object[] arguments);
+}
+----
+
+This section describes the behavior of the `invoke()` method.
+
+Calling `invoke()` invokes the target method on given `instance` of the target bean, passing given `arguments`.
+
+If the target method is `static`, the `instance` is ignored; by convention, it should be `null`.
+If the target method returns normally, its return value is returned, unless the target method is declared `void`, in which case `null` is returned.
+If the target method throws an exception, it is rethrown.
+The `invoke()` method does not declare any checked exception; implementations are supposed to use the "sneaky throw" idiom.
+// TODO maybe change that?
+
+If the target method is not `static` and `instance` is `null`, a `NullPointerException` is thrown.
+If the target method is not `static` and the `instance` is not <<invoker_assignability,assignable>> to the target bean, a `ClassCastException` is thrown.
+
+Correspondence between given `arguments` and declared parameters of the target method is positional: the Nth element of the `arguments` array is passed as the Nth argument to the target method.
+If the target method is a variable arity method, the last element of the `arguments` array corresponds to the variable arity parameter (and therefore must be an array).
+
+If the target method declares no parameter, `arguments` are ignored.
+If the target method declares any parameter and `arguments` is `null`, `NullPointerException` is thrown.
+If the `arguments` array has fewer elements than the number of parameters of the target method, `ArrayIndexOutOfBoundsException` is thrown.
+If the `arguments` array has more elements than the number of parameters of the target method, the excess elements are ignored.
+If some of the `arguments` is not <<invoker_assignability,assignable>> to the corresponding parameter of the target method, `ClassCastException` is thrown.
+
+If the declared type of a parameter of the target method is a primitive type, then:
+
+* if the corresponding argument is `null`, the default value of the primitive type is passed;
```

Comment:
Personally, I see this as unexpected and would rather have an exception thrown. The use cases where you'd deliberately use this "shortcut" instead of passing in the default yourself seem much more rare to me than cases in which you accidentally pass in `null` instead of actual non-default value.

Reply:
Yeah. Using default values makes certain things easier, but it is definitely a deviation of how things usually work. We can drop this, for sure.

Reply:
Agree we should fail on `null` for primitive parameters.